src/event_listener.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright 2025 Restate Software, Inc., Restate GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use std::ffi::CStr;
+use std::sync::Arc;
+
+use libc::c_void;
+
+use crate::ffi;
+use crate::ffi::{rocksdb_flushjobinfo_t, rocksdb_t};
+use crate::ffi_util::from_cstr;
+
+pub trait EventListenerExt {
+    fn add_event_listener<T>(&mut self, listener: Arc<T>)
+    where
+        T: EventListener + 'static + Send + Sync;
+}
+
+impl EventListenerExt for crate::Options {
+    fn add_event_listener<T>(&mut self, listener: Arc<T>)
+    where
+        T: EventListener + 'static + Send + Sync,
+    {
+        unsafe {
+            let cb = Box::new(EventListenerCallback::new(listener.clone()));
+            let cb_ptr = Box::into_raw(cb) as *mut c_void;
+
+            let event_listener = ffi::rocksdb_event_listener_create(
+                cb_ptr,
+                Some(EventListenerCallback::<T>::destructor),
+                Some(EventListenerCallback::<T>::name),
+            );
+
+            ffi::rocksdb_event_listener_set_on_flush_completed(
+                event_listener,
+                Some(EventListenerCallback::<T>::on_flush_completed),
+            );
+
+            ffi::rocksdb_options_add_event_listener(self.inner, event_listener);
+        }
+    }
+}
+
+/// EventListener defining a set of callbacks from RocksDB
+///
+/// The EventListener trait contains a set of callback functions that will
+/// be called when a specific RocksDB event happens such as flush. It can
+/// be used as a building block for developing custom features.
+///
+/// ## Important
+///
+/// Callback functions should not run for an extended period before they
+/// return as this will slow down RocksDB.
+///
+/// ## Threading
+///
+/// All EventListener callback will be called using the
+/// actual thread that involves in that specific event. For example, it
+/// is the RocksDB background flush thread that does the actual flush to
+/// call EventListener::OnFlushCompleted().
+///
+/// ## Locking
+///
+/// All EventListener callbacks are designed to be called without
+/// the current thread holding any DB mutex. This is to prevent potential
+/// deadlock and performance issue when using EventListener callbacks
+/// in a complex way.
+pub trait EventListener {
```

Comment:
You probably want to provide blanket implementations on this trait for Arc and Box as well

Reply:
I get it now! This is a much more idiomatic way of doin what I was hoping to accomplish.